macros.h

```diff
@@ -1,319 +1,459 @@
-      return NULL; \
-    } \
-    SN_STATUS_THROWS(napi_create_reference(env, cb, 1, &task->cb), "") \
-  } else { \
-    task->type = sn_async_task_promise; \
-    napi_create_promise(env, &task->deferred, &promise); \
-  }
+    task->req = (void *) req; \
+    js_value_t *promise; \
+    if (argc > cb_pos) { \
+      task->type = sn_async_task_callback; \
+      promise = NULL; \
+      err = js_create_reference(env, argv[cb_pos], 1, &task->cb); \
+      assert(err == 0); \
+    } else { \
+      task->type = sn_async_task_promise; \
+      err = js_create_promise(env, &task->deferred, &promise); \
+      assert(err == 0); \
+    }
 
 #define SN_ASYNC_COMPLETE(message) \
-  napi_value argv[1]; \
+  js_value_t *argv[1]; \
   switch (task->type) { \
   case sn_async_task_promise: { \
     if (task->code == 0) { \
-      napi_get_null(req->env, &argv[0]); \
-      napi_resolve_deferred(req->env, task->deferred, argv[0]); \
+      err = js_get_null(req->env, &argv[0]); \
+      assert(err == 0); \
+      err = js_resolve_deferred(req->env, task->deferred, argv[0]); \
+      assert(err == 0); \
     } else { \
-      napi_value err_msg; \
-      napi_create_string_utf8(req->env, #message, NAPI_AUTO_LENGTH, &err_msg); \
-      napi_create_error(req->env, NULL, err_msg, &argv[0]); \
-      napi_reject_deferred(req->env, task->deferred, argv[0]); \
+      js_value_t *err_msg; \
+      assert(err == 0); \
+      err = js_create_string_utf8(req->env, (const utf8_t *) #message, SN_LIT_LENGTH(message), &err_msg); \
+      assert(err == 0); \
+      err = js_create_error(req->env, NULL, err_msg, &argv[0]); \
+      assert(err == 0); \
+      err = js_reject_deferred(req->env, task->deferred, argv[0]); \
+      assert(err == 0); \
     } \
     task->deferred = NULL; \
     break; \
   } \
   case sn_async_task_callback: { \
     SN_CALLBACK_CHECK_FOR_ERROR(#message) \
-    napi_value callback; \
-    napi_get_reference_value(req->env, task->cb, &callback); \
-    napi_value return_val; \
+    js_value_t *callback; \
+    err = js_get_reference_value(req->env, task->cb, &callback); \
+    assert(err == 0); \
+    js_value_t *return_val; \
     SN_CALL_FUNCTION(req->env, global, callback, 1, argv, &return_val) \
-    napi_close_handle_scope(req->env, scope); \
-    napi_delete_reference(req->env, task->cb); \
+    err = js_delete_reference(req->env, task->cb); \
+    assert(err == 0); \
     break; \
   } \
   }
+
+#define SN_TYPEDARRAY_VIEW(var) \
+  void *var##_data; \
+  size_t var##_length; \
+  js_typedarray_type_t var##_type; \
+  js_typedarray_view_t *var##_view; \
+  \
+  err = js_get_typedarray_view(env, var, &var##_type, &var##_data, &var##_length, &var##_view); \
+  assert(err == 0); \
+  \
+  uint8_t var##_width = typedarray_width(var##_type); \
+  assert(var##_width != 0 && "Unexpected TypedArray type"); \
+  \
+  size_t var##_size = var##_length * var##_width;
+
+#define SN_TYPEDARRAY_VIEW_OPT(var) \
```

Comment:
the other optional methods use the convention `SN_OPT_TYPEDARRAY_VIEW`, this does seem cleaner though, but maybe we should change all the other methods then?

Reply:
seeing this a bit late but yes, i think you're right, but id' rather just conform for now and make one more pull :)

Reply:
resolved?

Reply:
yes , renamed `JS_TYPEDARRAY_VIEW_OPT` => `JS_OPT_TYPEDARRAY_VIEW`

Reply:
nice! is that all of @chm-diederichs comments then?

Reply:
(we agreed yes off screen)